crates/sui-types/src/dynamic_field.rs

```diff
@@ -368,39 +369,257 @@ where
+    }
 }
 
-/// Get a dynamic field from the store with a generic key type. The key type may
-/// only have phantom type parameters, because it must have a fixed serialization
-/// format.
-pub fn get_dynamic_field_from_store_generic<K, V>(
-    object_store: &dyn ObjectStore,
-    parent_id: ObjectID,
-    key: &K,
-    key_type_params: &[TypeTag],
-) -> Result<V, SuiError>
+/// A DynamicField is a `Field<K, V>` object, that is, the value that would be stored in the field object.
+pub struct DynamicField<K, V>(Field<K, V>, TypeTag);
+
+impl<K, V> DynamicField<K, V>
 where
-    K: MoveTypeTagTraitGeneric + Serialize + DeserializeOwned + fmt::Debug,
-    V: Serialize + DeserializeOwned,
+    K: Serialize,
+    V: Serialize,
+{
+    /// Convert the internal `Field<K, V>` object into a Move object.
+    /// Use this to create a Move object in memory without reading one from the store.
+    ///
+    /// IMPORTANT: Do not call except in tests to avoid possible conservation bugs.
+    /// (For instance, you can simply create a Move object that contains minted SUI. If
+    /// you then write this object to the db, it will break conservation.)
+    pub fn into_move_object_unsafe_for_testing(
+        self,
+        version: SequenceNumber,
+    ) -> Result<MoveObject, SuiError> {
+        let field = self.0;
+        let type_tag = self.1;
+        let TypeTag::Struct(struct_tag) = type_tag else {
+            unreachable!()
+        };
+        // TODO(address-balances): more efficient type repr
+        let move_object_type = MoveObjectType::from(*struct_tag);
+
+        let field_bytes =
+            bcs::to_bytes(&field).map_err(|e| SuiError::ObjectSerializationError {
+                error: e.to_string(),
+            })?;
+        Ok(unsafe {
+            MoveObject::new_from_execution_with_limit(
+                move_object_type,
+                false, // A dynamic field is never transferable, public or otherwise.
+                version,
+                field_bytes,
+                512,
+            )
+        }?)
+    }
+
+    /// Get the internal `Field<K, V>` object.
+    pub fn into_inner(self) -> Field<K, V> {
+        self.0
+    }
+}
+
+/// A DynamicFieldID contains the material needed to load an a dynamic field from
+/// the store.
+///
+/// Can be obtained from a DynamicFieldKey, or created directly if you know the
+/// parent and child IDs but do not know the key.
+pub struct DynamicFieldID<K: Serialize>(
+    pub ObjectID, // parent
+    pub ObjectID, // child
+    std::marker::PhantomData<K>,
+);
+
+impl<K> DynamicFieldID<K>
+where
+    K: Serialize + std::fmt::Debug,
+{
+    /// Create a DynamicFieldID from a parent and child ID.
+    pub fn new(parent: ObjectID, id: ObjectID) -> Self {
+        Self(parent, id, std::marker::PhantomData)
+    }
+
+    /// Load the field object from the store.
+    pub fn load_object(self, object_store: &dyn ObjectStore) -> Option<DynamicFieldObject<K>> {
```

Comment:
I still don't understand this one. So this is loading the parent object?

Reply:
I rewrote the pre-existing `get_dynamic_field_object_from_store` and `get_dynamic_field_from_store` to use this code so that it is covered by existing tests, and fixed the bug